src/main/java/apoc/file/Neo4jFileStream.java

```diff
@@ -0,0 +1,89 @@
+import apoc.util.FileUtils;
+import org.neo4j.procedure.Description;
+import org.neo4j.procedure.Name;
+import org.neo4j.procedure.Procedure;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicLong;
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.stream.Stream;
+
+public class Neo4jFileStream {
+    public class FileEntry implements Comparable<FileEntry> {
+        public long lineNo;
+        public String line;
+        public String path;
+
+        public FileEntry(long lineNumber, String data, String path) {
+            this.lineNo = lineNumber;
+            this.line = data;
+            this.path = path;
+        }
+
+        public int compareTo(FileEntry o) {
+            if (this.lineNo < o.lineNo) {
+                return -1;
+            } else if(this.lineNo > o.lineNo) {
+                return 1;
+            }
+
+            return 0;
+        }
+    }
+
+    @Procedure
+    @Description( "apoc.file.stream(path, config) - retrieve system or log file contents" )
+    public Stream<FileEntry> stream(
+            @Name("path") String path,
+            @Name(value = "config",defaultValue = "{}") Map<String, Object> config) {
+        String neo4jHome = ApocConfiguration.allConfig.get("unsupported.dbms.directories.neo4j_home");
+        // String logDir = ApocConfiguration.allConfig.get("dbms.directories.logs");
+
+        // Prepend neo4jHome if it's a relative path, and use the user's path otherwise.
+        String fullPath = path.startsWith(File.separator) ? path : (neo4jHome + File.separator + path);
+        String canonicalPath = null;
+        File f = new File(fullPath);
+
+        try {
+            canonicalPath = f.getCanonicalPath();
+            // Open question on whether this should be permitted or not.  It is permitted by other APOC utilities,
+            // so for now it's permitted here with a warning.
+            if (!canonicalPath.equals(f.getAbsolutePath()) && !canonicalPath.contains(neo4jHome)) {
+                System.err.println("Warning: user attempting to access file outside of neo4jHome " +
+                        canonicalPath);
+            }
+        } catch (IOException ioe) {
+            throw new RuntimeException("Unable to resolve basic file canonical path", ioe);
+        }
+
+        // Throws error if APOC is not configured to permit local file reads.
+        FileUtils.checkReadAllowed(f.toURI().toString());
+
+        try {
+            Stream<String> stream = Files.lines(Paths.get(fullPath));
+            final AtomicLong lineNumber = new AtomicLong(0);
+            final String p = canonicalPath;
+
+            Stream<FileEntry> entries = stream.map(line -> new FileEntry(lineNumber.getAndIncrement(), line, p));
+
+            // Useful for tailing logfiles.
+            if(config.containsKey("last")) {
```

Comment:
I've investigated.  Time range filters, no.  Because the format of logfiles aren't consistent, and because parsing would be required.  I think to make it simple to stream lines of files, but not to go further than this, otherwise a bunch of additional procedures are needed specialized for each file.  Because these files can include stacktraces in them, the parsing gets quite tricky quickly if what you want to do is give a stream of "log entries" rather than a stream of lines.   They're not the same unfortunately.   This is intended as a utility just to provide access to this data but it's left to the user how to access.

Reply:
Makes sense.